lib/project-data.ts

```diff
@@ -47,46 +53,121 @@ export class ProjectData implements IProjectData {
+		// If no project found, projectDir should be null
+		let nsData: any = null;
+		let nsConfig: any = null;
+		let packageJsonData: any = null;
+
+		try {
+			packageJsonData = parseJson(packageJsonContent);
+			nsData = packageJsonData[this.$staticConfig.CLIENT_NAME_KEY_IN_PROJECT_FILE];
+		} catch (err) {
+			this.$errors.failWithoutHelp(`The project file ${this.projectFilePath} is corrupted. ${EOL}` +
+				`Consider restoring an earlier version from your source control or backup.${EOL}` +
+				`Additional technical info: ${err.toString()}`);
 		}
 
+		try {
+			nsConfig = nsconfigContent ? parseJson(nsconfigContent) : null;
+		} catch (err) {
+			this.$errors.failWithoutHelp(`The NativeScript configuration file ${constants.CONFIG_NS_FILE_NAME} is corrupted. ${EOL}` +
+				`Consider restoring an earlier version from your source control or backup.${EOL}` +
+				`Additional technical info: ${err.toString()}`);
+		}
+
+		if (nsData) {
+			this.projectDir = projectDir;
+			this.projectName = this.$projectHelper.sanitizeName(path.basename(projectDir));
+			this.platformsDir = path.join(projectDir, constants.PLATFORMS_DIR_NAME);
+			this.projectFilePath = projectFilePath;
+			this.projectId = nsData.id;
+			this.dependencies = packageJsonData.dependencies;
+			this.devDependencies = packageJsonData.devDependencies;
+			this.projectType = this.getProjectType();
+			this.nsConfig = nsConfig;
+
+			return;
+		}
+
+		this.errorInvalidProject(projectDir);
+	}
+
+	private errorInvalidProject(projectDir: string) {
 		const currentDir = path.resolve(".");
 		this.$logger.trace(`Unable to find project. projectDir: ${projectDir}, options.path: ${this.$options.path}, ${currentDir}`);
 
 		// This is the case when no project file found
 		this.$errors.fail("No project found at or above '%s' and neither was a --path specified.", projectDir || this.$options.path || currentDir);
 	}
 
+	private getProjectFilePath(projectDir: string): string {
+		return path.join(projectDir, this.$staticConfig.PROJECT_FILE_NAME);
+	}
+
+	public getAppResourcesDirectoryPath(projectDir?: string): string {
+		if (!projectDir) {
+			projectDir = this.projectDir;
+		}
+
+		return path.resolve(projectDir, this.getAppResourcesRelativeDirectoryPath());
+	}
+
+	public getAppResourcesRelativeDirectoryPath(): string {
+		if (this.nsConfig && this.nsConfig[constants.CONFIG_NS_APP_RESOURCES_ENTRY]) {
+			return this.nsConfig[constants.CONFIG_NS_APP_RESOURCES_ENTRY];
+		}
+
+		return path.join(this.getAppDirectoryRelativePath(), constants.APP_RESOURCES_FOLDER_NAME);
+	}
+
+	public getAppDirectoryPath(projectDir?: string): string {
+		if (!projectDir) {
+			projectDir = this.projectDir;
+		}
+
+		return path.resolve(projectDir, this.getAppDirectoryRelativePath());
+	}
+
+	public getAppDirectoryRelativePath(): string {
+		if (this.nsConfig && this.nsConfig[constants.CONFIG_NS_APP_ENTRY]) {
+			return this.nsConfig[constants.CONFIG_NS_APP_ENTRY];
```

Comment:
what will happen in case the path in the nsconfig.json is not relative?

Reply:
I think we should add validation for this as supporting paths not relative to the project will be very hard in cloud builds. Maybe we should also change the key inside `nsconfig.json` to appRelativePath or something in this lines.